File: tech-studio-projects/machine-learning/credit-card-fraud-detection-simulation/model_training.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""model_trainig.py

The kaggle dataset (data/creditcard.csv) is used to train a ML model
for detecting the fraud. A Random Forest Classifier is used here to, because
it is robust and easy to implement.
"""

import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score
import joblib


def train_model() -> None:
    """Train and save the model."""
    # Load the dataset
    df = pd.read_csv('data/creditcard.csv')

    # Split features and labels
    X = df.drop(['Class'], axis=1)
    y = df['Class']

    # Split into training and test sets
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    # Standardize the feature values
    scaler = StandardScaler()
    X_train_scaled = scaler.fit_transform(X_train)
    X_test_scaled = scaler.transform(X_test)

    # Train a Random Forest Classifier
    model = RandomForestClassifier(n_estimators=100, random_state=42)
    model.fit(X_train_scaled, y_train)

    # Evaluate the model
    y_pred = model.predict(X_test_scaled)
    accuracy = accuracy_score(y_test, y_pred)
    print(f'Model Accuracy: {accuracy}')

    # Save the model and scaler for use in the consumer
    joblib.dump(model, 'models/fraud_detection_model.pkl')
    joblib.dump(scaler, 'models/scaler.pkl')

    return None

File: tech-studio-projects/machine-learning/credit-card-fraud-detection-simulation/java/ModelTraining.java
import weka.classifiers.trees.RandomForest;
import weka.core.converters.ConverterUtils.DataSource;
import weka.core.Instances;
import weka.filters.Filter;
import weka.filters.unsupervised.attribute.Standardize;
import java.io.FileOutputStream;
import java.io.ObjectOutputStream;

public class ModelTraining {

    public static void main(String[] args) {
        try {
            // Load the dataset
            DataSource source = new DataSource("data/creditcard.arff"); // Use ARFF file format for Weka
            Instances data = source.getDataSet();

            // Ensure class attribute is the last column
            if (data.classIndex() == -1) {
                data.setClassIndex(data.numAttributes() - 1);
            }

            // Split features and labels
            Instances features = new Instances(data);
            features.deleteAttributeAt(data.classIndex());  // Remove the class attribute

            // Standardize the features
            Standardize standardize = new Standardize();
            standardize.setInputFormat(features);
            Instances standardizedFeatures = Filter.useFilter(features, standardize);

            // Add the class attribute back to the standardized dataset
            standardizedFeatures.insertAttributeAt(data.classAttribute(), standardizedFeatures.numAttributes());

            // Split the dataset into training (80%) and test (20%) sets
            int trainSize = (int) Math.round(data.numInstances() * 0.8);
            int testSize = data.numInstances() - trainSize;
            Instances trainData = new Instances(standardizedFeatures, 0, trainSize);
            Instances testData = new Instances(standardizedFeatures, trainSize, testSize);

            // Train a Random Forest Classifier
            RandomForest rf = new RandomForest();
            rf.setNumTrees(100);
            rf.buildClassifier(trainData);

            // Evaluate the model
            double accuracy = evaluateModel(rf, testData);
            System.out.println("Model Accuracy: " + accuracy);

            // Save the model and scaler
            saveModel(rf, "models/fraud_detection_model.model");
            saveStandardizer(standardize, "models/scaler.model");

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static double evaluateModel(RandomForest model, Instances testData) throws Exception {
        int correct = 0;
        for (int i = 0; i < testData.numInstances(); i++) {
            double actualClass = testData.instance(i).classValue();
            double predictedClass = model.classifyInstance(testData.instance(i));
            if (actualClass == predictedClass) {
                correct++;
            }
        }
        return (double) correct / testData.numInstances();
    }

    public static void saveModel(RandomForest model, String modelPath) {
        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(modelPath))) {
            oos.writeObject(model);
            System.out.println("Model saved to " + modelPath);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void saveStandardizer(Standardize standardize, String scalerPath) {
        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(scalerPath))) {
            oos.writeObject(standardize);
            System.out.println("Scaler saved to " + scalerPath);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
